src/coreclr/jit/helperexpansion.cpp

```diff
@@ -2712,3 +2712,166 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
+
+    if (!doesMethodHaveStackAllocatedArray())
+    {
+        // The method being compiled doesn't have any stack allocated arrays.
+        return result;
+    }
+
+    // Find allocation sites, and transform them into initializations of the
+    // array method table and length, and replace the allocation call with
+    // the address of the local array.
+    //
+    bool modified = false;
+
+    for (BasicBlock* const block : Blocks())
+    {
+        for (Statement* const stmt : block->Statements())
+        {
+            if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+            {
+                continue;
+            }
+
+            for (GenTree* const tree : stmt->TreeList())
+            {
+                if (!tree->IsCall())
+                {
+                    continue;
+                }
+
+                if (fgExpandStackArrayAllocation(block, stmt, tree->AsCall()))
+                {
+                    // If we expand, we split the statement's tree
+                    // so will be done with this statment.
+                    //
+                    modified = true;
+                    break;
+                }
+            }
+        }
+    }
+
+    // we cant assert(modified) here as array allocation sites may
+    // have been unreachable or dead-coded.
+    //
+    return modified ? PhaseStatus::MODIFIED_EVERYTHING : PhaseStatus::MODIFIED_NOTHING;
+}
+
+//------------------------------------------------------------------------------
+// fgExpandStackArrayAllocation: expand new array helpers for stack allocated arrays
+//
+// Arguments:
+//    block -  block containing the helper call to expand
+//    stmt   - Statement containing the helper call
+//    call   - The helper call
+//
+// Returns:
+//    true if a runtime lookup was found and expanded.
+//
+bool Compiler::fgExpandStackArrayAllocation(BasicBlock* block, Statement* stmt, GenTreeCall* call)
+{
+    if (!call->IsHelperCall())
+    {
+        return false;
+    }
+
+    const CorInfoHelpFunc helper         = eeGetHelperNum(call->gtCallMethHnd);
+    int                   lengthArgIndex = -1;
+
+    switch (helper)
+    {
+        case CORINFO_HELP_NEWARR_1_DIRECT:
+        case CORINFO_HELP_NEWARR_1_VC:
+        case CORINFO_HELP_NEWARR_1_OBJ:
+        case CORINFO_HELP_NEWARR_1_ALIGN8:
+            lengthArgIndex = 1;
+            break;
+
+        case CORINFO_HELP_READYTORUN_NEWARR_1:
+            lengthArgIndex = 0;
+            break;
+
+        default:
+            return false;
+    }
+
+    if ((call->gtCallMoreFlags & GTF_CALL_M_STACK_ARRAY) == 0)
+    {
+        return false;
+    }
+
+    CallArg* const             stackLocalAddressArg  = call->gtArgs.FindWellKnownArg(WellKnownArg::StackArrayLocal);
+    GenTreeLclVarCommon* const stackLocalAddressNode = stackLocalAddressArg->GetNode()->AsLclVarCommon();
+    const unsigned             lclNum                = stackLocalAddressNode->GetLclNum();
```

Comment:
What guarantees that this `LCL_ADDR` makes it through the JIT? This looks pretty fragile, I do not think we should be making assumptions on the shape of this argument node.

Reply:
Seems like the code below can work just fine without making assumptions about the shape (potentially just optimizing for the `LCL_ADDR` shape)

Reply:
So you are suggesting just using indirs to store the method table/length rather than fields? That is, transform
```
temp = newarray (mt; len, localarg)
```
into
```
temp = localarg
*temp = mt
*(temp + 8) = len
```
instead of what we're doing now?
```
*(&L) = mt
*(&L + 8) = len
temp = &L
```

Reply:
Yeah, that's what I was thinking.

(As a side note you are guaranteed after `gtSplitTree` that `localarg` is an invariant node, so you can clone it directly via `gtCloneExpr` without introducing any temp yourself)